File: core/src/com/lagecompany/infinity/anim/Animator.java
package com.lagecompany.infinity.anim;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.files.FileHandle;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.Animation;
import com.badlogic.gdx.graphics.g2d.Animation.PlayMode;
import com.badlogic.gdx.graphics.g2d.Batch;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.badlogic.gdx.scenes.scene2d.Actor;
import com.badlogic.gdx.utils.Json;
import com.lagecompany.infinity.anim.AnimationData.AnimationDataRegion;

public class Animator extends Actor {
	private static final String LOG_TAG = Animator.class.getSimpleName();

	private final Map<String, Animation<TextureRegion>> animationMap;
	private String currentAnimationName;
	private Animation<TextureRegion> currentAnimation;
	private float stateTime;

	public Animator() {
		this.animationMap = new HashMap<>();
	}

	public String addAnimation(String path) {
		try {
			path = "anim/" + path + ".anim";
			FileHandle fileHandle = Gdx.files.internal(path);
			Json json = new Json();
			AnimationData data = json.fromJson(AnimationData.class, fileHandle);

			if (data == null) {
				Gdx.app.error(LOG_TAG, "Failed to parse animation file " + path);
			} else {
				Texture texture = new Texture(Gdx.files.internal(data.getTexturePath()));

				List<TextureRegion> regions = new ArrayList<>();
				for (AnimationDataRegion dataRegion : data.getRegions()) {
					regions.add(new TextureRegion(texture, dataRegion.getX(), dataRegion.getY(), dataRegion.getWidth(),
							dataRegion.getHeight()));
				}

				Animation<TextureRegion> animation = new Animation<TextureRegion>(data.getFrameTime(),
						regions.toArray(new TextureRegion[regions.size()]));
				animation.setPlayMode(PlayMode.LOOP);
				animationMap.put(data.getName(), animation);
				return data.getName();
			}
		} catch (Exception e) {
			Gdx.app.error(LOG_TAG, "Failed to load animation at path " + path, e);
		}
		return null;
	}

	public boolean playAnimation(String name) {
		Animation<TextureRegion> animation = animationMap.get(name);

		if (animation == null)
			return false;

		currentAnimationName = name;
		currentAnimation = animation;
		stateTime = 0f;
		return true;
	}

	public String getCurrentAnimationName() {
		return currentAnimationName;
	}

	public void stopAnimation() {
		currentAnimation = null;
	}

	@Override
	public void draw(Batch batch, float parentAlpha) {
		if (currentAnimation == null)
			return;

		TextureRegion region = currentAnimation.getKeyFrame(stateTime);
		batch.draw(region, getX(), getY());
	}

	@Override
	public void act(float delta) {
		super.act(delta);
		stateTime += delta;
	}

}

File: core/src/main/kotlin/com/lagecompany/infinty/animation/Animator.kt
package com.lagecompany.infinty.animation

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.graphics.g2d.Animation
import com.badlogic.gdx.graphics.g2d.Animation.PlayMode
import com.badlogic.gdx.graphics.g2d.Batch
import com.badlogic.gdx.graphics.g2d.TextureRegion
import com.badlogic.gdx.scenes.scene2d.Actor
import com.badlogic.gdx.utils.Json

data class AnimationRegion(val x: Int = 0, val y: Int = 0, val width: Int = 0, val height: Int = 0)
data class AnimationData(
        val name: String = "",
        val texturePath: String = "",
        val frameTime: Float = 0.1f,
        val flip: Boolean = false,
        val loop: Boolean = true
) {
    val regions = mutableListOf<AnimationRegion>()

    fun addRegion(x: Int, y: Int, width: Int, height: Int) {
        regions.add(AnimationRegion(x, y, width, height))
    }
}

private const val LOG_TAG = "ANIMATOR"

class Animator : Actor() {
    private val animationMap = mutableMapOf<String, Animation<TextureRegion>>()

    private lateinit var currentAnimation: Animation<TextureRegion>

    var currentAnimationName = ""
        private set

    private val isPlaying: Boolean
        get() = currentAnimationName.isNotEmpty()

    private var stateTime = 0f

    fun addAnimation(path: String) {
        try {
            val fileHandle = Gdx.files.internal("anim/$path.anim")
            val data = Json().fromJson(AnimationData::class.java, fileHandle)

            if (data != null) {
                val texture = Texture(Gdx.files.internal(data.texturePath))

                var regions = mutableListOf<TextureRegion>()
                for (dataRegion in data.regions) {
                    val region = TextureRegion(texture, dataRegion.x, dataRegion.y, dataRegion.width, dataRegion.height)
                    if (data.flip)
                        region.flip(true, false)
                    regions.add(region)
                }

                var animation = Animation<TextureRegion>(data.frameTime, *regions.toTypedArray())
                animation.playMode = if (data.loop) PlayMode.LOOP else PlayMode.NORMAL
                animationMap.put(data.name, animation)
            } else {
                Gdx.app.error(LOG_TAG, "Failed to parse animation file $path")
            }
        } catch (e: Exception) {
            Gdx.app.error(LOG_TAG, "Failed to load animation at path $path", e)
        }
    }

    fun playAnimation(name: String): Boolean {
        if (name.equals(currentAnimationName, ignoreCase = true))
            return false

        val animation = animationMap.get(name) ?: return false

        currentAnimationName = name
        currentAnimation = animation
        stateTime = 0f

        return true
    }

    fun stopAnimation() {
        currentAnimationName = ""
    }

    override fun draw(batch: Batch?, parentAlpha: Float) {
        if (!isPlaying)
            return

        val region = currentAnimation.getKeyFrame(stateTime)
        batch?.draw(region, x, y, region.regionX.toFloat(), region.regionY.toFloat(), region.regionWidth.toFloat(),
                region.regionHeight.toFloat(), scaleX, scaleY, rotation)
    }

    override fun act(delta: Float) {
        super.act(delta)
        stateTime += delta
    }
}